File: Tools/BuildTool/Source/Toolchains/IToolchain.cs
using System;
using System.Collections.Generic;
using System.Diagnostics;
using System.IO;
using System.Linq;
using System.Reflection;

public enum BuildType
{
    Debug,
    Development,
    Release
}

public class BuildPlatform
{
    public String platformName;
    public List<String> preprocessorDefines = new List<string>();
    public List<String> linkerLibraries = new List<string>();
}


public abstract class IToolchain
{
    private static List<IToolchain> _toolchains = null;

    public static List<IToolchain> GetToolchains()
    {
        if (_toolchains == null)
        {
            _toolchains = new List<IToolchain>();

            var types = typeof(IToolchain).Assembly.GetTypes().Where(t => t.IsSubclassOf(typeof(IToolchain)));
            foreach (var type in types)
            {
                if (!type.IsAbstract)
                {
                    var instanceProperty = type.GetProperty("Instance",
                        BindingFlags.Public | BindingFlags.Static | BindingFlags.FlattenHierarchy);
                    if (instanceProperty != null)
                    {

                        try
                        {
                            IToolchain toolchain = (IToolchain) instanceProperty.GetValue(null, null);
                            _toolchains.Add(toolchain);
                        }
                        catch (Exception)
                        {
                        }
                    }
                }
            }
        }

        return _toolchains;
    }

    protected static int ExecuteCommand(string command, out string output)
    {
        Process process = new Process();
        process.StartInfo.UseShellExecute = false;
        process.StartInfo.RedirectStandardOutput = true;
        process.StartInfo.RedirectStandardError = true;
        process.StartInfo.FileName = "cmd.exe";
        process.StartInfo.Arguments = "/C" + command;
        process.Start();

        output = process.StandardOutput.ReadToEnd();
        output += process.StandardError.ReadToEnd();
        process.WaitForExit();

        return process.ExitCode;
    }

    protected static int ExecuteCommand(string command)
    {
        string discardLog = "";
        return ExecuteCommand(command, out discardLog);
    }


    public abstract List<BuildPlatform> GetPlatforms();

    public abstract bool PrepareModuleForBuild(ModuleManifest manifest, BuildPlatform target);

    public abstract bool BuildSource(ModuleManifest manifest, BuildPlatform target, string sourceFile, List<string> includePaths, List<string> preprocessorDefines);

    public abstract bool LinkLibrary(ModuleManifest manifest, BuildPlatform target, List<string> sourceFiles);

    public abstract bool LinkExecutable(ModuleManifest manifest, BuildPlatform target, List<string> sourceFiles);

    public string GetObjDirectory(ModuleManifest manifest, BuildPlatform target)
    {
        return manifest.buildRoot + "\\Obj\\" + target.platformName.ToLower().Replace(' ', '_');
    }

    public string GetObjPath(ModuleManifest manifest, BuildPlatform target, string sourceFile)
    {
        return GetObjDirectory(manifest, target) + "\\" + Path.GetFileNameWithoutExtension(sourceFile) + ".obj";
    }

    public abstract string GetObjExtension();

    public string GetLibDirectory(ModuleManifest manifest)
    {
        return manifest.buildRoot + "\\Lib";
    }

    public string GetLibPath(ModuleManifest manifest, BuildPlatform target)
    {
        return GetLibDirectory(manifest) + "\\" + target.platformName.ToLower().Replace(' ', '_') + "." + GetLibExtension();
    }

    public string GetExePath(ModuleManifest manifest, BuildPlatform target)
    {
        return GetLibDirectory(manifest) + "\\" + manifest.name + "_" + target.platformName.ToLower().Replace(' ', '_') + "." + GetExeExtension();
    }

    public abstract string GetLibExtension();

    public abstract string GetExeExtension();
}

public abstract class IToolchainInterface<T> : IToolchain where T : IToolchain, new()
{
    public static T Instance
    {
        get
        {
            if (_instance == null)
            {
                _instance = new T();
            }

            return _instance;
        }
    }

    private static T _instance = null;
}

File: Tools/BuildTool/Source/Toolchains/IToolchain.cpp
#include "IToolchain.hpp"
#include <Utility.hpp>
#include <Manifests/ModuleManifest.hpp>
#include <Manifests/ProjectManifest.hpp>
#include "Command.hpp"
#include "ToolchainMSVC.hpp"
#include "ToolchainEmscripten.hpp"

std::vector<std::string> IToolchain::GetPreprocessorDefines(ModuleManifest* manifest, BuildPlatform* platform, BuildType configuration)
{
    std::vector<std::string> defines;

    std::set<std::string> platformDefines;
    for (IToolchain* toolchain : IToolchain::GetToolchains())
    {
        for (BuildPlatform* toolchainPlatform : toolchain->GetPlatforms())
        {
            for (std::string& define : toolchainPlatform->preprocessorDefines)
            {
                platformDefines.insert(define);
            }
        }
    }

    for (std::string define : platformDefines)
    {
        bool found = std::find(platform->preprocessorDefines.begin(), platform->preprocessorDefines.end(), define) != platform->preprocessorDefines.end();
        defines.push_back(define + "=" + (found ? '1' : '0'));
    }

    defines.push_back(std::string("FARO_DEBUG=") + (configuration == Debug ? '1' : '0'));
    defines.push_back(std::string("FARO_DEVELOPMENT=") + (configuration == Development ? '1' : '0'));
    defines.push_back(std::string("FARO_RELEASE=") + (configuration == Release ? '1' : '0'));

    if (manifest != nullptr)
    {
        std::vector<std::string> moduleDefines = manifest->GetModuleDefines();
        for (std::string& define : moduleDefines)
        {
            defines.push_back(define);
        }
    }

    return defines;
}

std::set<ModuleManifest*> IToolchain::GetAllModuleDependencies(ModuleManifest& topModule)
{
    std::vector<std::filesystem::path> toProcess = { topModule.moduleDependencies };
    std::set<ModuleManifest*> dependencies;

    while (!toProcess.empty())
    {
        std::filesystem::path dependencyPath = toProcess[0];
        toProcess.erase(toProcess.begin());

        ModuleManifest* dependency = ModuleManifest::GetLoadedModule(dependencyPath);

        if (dependencies.find(dependency) == dependencies.end())
        {
            dependencies.insert(dependency);
            for (std::filesystem::path childDependencyPath : dependency->moduleDependencies)
            {
                toProcess.push_back(childDependencyPath);
            }
        }
    }

    return dependencies;
}

int IToolchain::ExecuteCommand(std::string command, std::string& output)
{
    return Utility::ExecuteCommand(command, output);
}

int IToolchain::ExecuteCommand(std::string command)
{
    std::string discardLog = "";
    return ExecuteCommand(command, discardLog);
}

std::vector<IToolchain*> IToolchain::GetToolchains()
{
    return { &ToolchainMSVC::Instance, &ToolchainEmscripten::Instance };
}

std::filesystem::path IToolchain::GetObjDirectory(ModuleManifest& manifest, BuildPlatform* target, BuildType configuration)
{
    std::string platformName = target->platformName + " " + BuildTypeNames[configuration];
    platformName = Utility::ToLower(platformName);
    std::replace(platformName.begin(), platformName.end(), ' ', '_');
    return manifest.faroDirectory / "Obj" / platformName;
}

std::filesystem::path IToolchain::GetObjPath(ModuleManifest& manifest, BuildPlatform* target, BuildType configuration, std::filesystem::path sourceFile)
{
    return GetObjDirectory(manifest, target, configuration) / sourceFile.filename().replace_extension(GetObjExtension());
}

std::filesystem::path IToolchain::GetBinDirectory(ModuleManifest& manifest)
{
    return manifest.faroDirectory / "Bin";
}

std::filesystem::path IToolchain::GetLibPath(ModuleManifest& manifest, BuildPlatform* target, BuildType configuration)
{
    std::string platformName = target->platformName + " " + BuildTypeNames[configuration];
    platformName = Utility::ToLower(platformName);
    std::replace(platformName.begin(), platformName.end(), ' ', '_');

    return GetBinDirectory(manifest) / (platformName + "." + GetLibExtension());
}

std::filesystem::path IToolchain::GetExePath(ModuleManifest& manifest, BuildPlatform* target, BuildType configuration)
{
    std::string platformName = target->platformName + " " + BuildTypeNames[configuration];
    platformName = Utility::ToLower(platformName);
    std::replace(platformName.begin(), platformName.end(), ' ', '_');

    return GetBinDirectory(manifest) / (manifest.name + "_" + platformName + "." + GetExeExtension());
}

std::filesystem::path IToolchain::GetBinaryPath(ModuleManifest& manifest, BuildPlatform* target, BuildType configuration)
{
    if (manifest.type == MT_Library)
    {
        return GetLibPath(manifest, target, configuration);
    }
    else if (manifest.type == MT_Executable)
    {
        return GetExePath(manifest, target, configuration);
    }
    return {};
}
